src/environment.ts

```diff
@@ -30,90 +19,33 @@ import {
-    WeakMapSet(cachedReflectiveIntrinsicsMap, global, reflectiveIntrinsics);
-    // remapping intrinsics that are realm's agnostic
-    for (let i = 0, len = ReflectiveIntrinsicObjectNames.length; i < len; i += 1) {
-        const name = ReflectiveIntrinsicObjectNames[i];
-        reflectiveIntrinsics[name] = global[name];
-    }
-    return reflectiveIntrinsics;
-}
-
-// caching from the blue realm right away to avoid picking up modified entries
-getReflectiveIntrinsics(globalThis);
-
-export class SecureEnvironment implements MembraneBroker {
+export class MembraneBroker {
     // map from red to blue references
-    redMap: WeakMap<RedFunction | RedObject, RedProxyTarget | BlueProxy> = WeakMapCreate();
+    redMap: WeakMap<RedFunction | RedObject, RedProxyTarget | BlueProxy>;
     // map from blue to red references
-    blueMap: WeakMap<BlueFunction | BlueObject, RedProxy | BlueProxyTarget> = WeakMapCreate();
+    blueMap: WeakMap<BlueFunction | BlueObject, RedProxy | BlueProxyTarget>;
     // blue object distortion map
     distortionMap: DistortionMap;
 
-    constructor(options: SecureEnvironmentOptions) {
-        if (isUndefined(options)) {
-            throw ErrorCreate(`Missing SecureEnvironmentOptions options bag.`);
-        }
-        const { blueGlobalThis, redGlobalThis, distortionMap } = options;
-        this.distortionMap = WeakMapCreate();
-        // validating distortion entries
-        distortionMap?.forEach((value, key) => {
-            if (typeof key !== typeof value) {
-                throw ErrorCreate(`Invalid distortion ${value}.`);
-            }
-            WeakMapSet(this.distortionMap, key, value);
-        });
-        // getting proxy factories ready per environment so we can produce
-        // the proper errors without leaking instances into a sandbox
-        const redEnvFactory = redGlobalThis.eval(`(${serializedRedEnvSourceText})`);
-        const blueHooks: MarshalHooks = {
-            apply(target: BlueFunction, thisArgument: BlueValue, argumentsList: ArrayLike<BlueValue>): BlueValue {
-                return apply(target, thisArgument, argumentsList);
-            },
-            construct(target: BlueConstructor, argumentsList: ArrayLike<BlueValue>, newTarget?: any): BlueValue {
-                return construct(target, argumentsList, newTarget);
-            },
-        };
-        this.getRedValue = redEnvFactory(this, blueHooks);
+    constructor(
+        registry: SandboxRegistry,
+        blueProxyFactory: (red: RedValue) => BlueValue,
+        redProxyFactory: (blue: BlueValue) => RedValue
+    ) {
+        const { redMap, blueMap, distortionMap } = registry;
+        this.redMap = redMap;
+        this.blueMap = blueMap;
+        this.distortionMap = distortionMap;
+
         this.getBlueValue = blueProxyFactory(this);
-        // remapping intrinsics that are realm's agnostic
-        const blueIntrinsics = getReflectiveIntrinsics(blueGlobalThis);
```

Comment:
this is done from outside now (caller is responsible for that, instead of the broker itself)